### PR TITLE
PID Controller

### PR DESCRIPTION
**Pull request by stevendes**
**WORK IN PROGRESS - UPDATE 2**

Working code, the variables are now more in order to python standards, still have some cleaning and simplifications to do in the code.

It works by using a **callback** function to get the actual position of the robot, and starting from that it sets the new goals to draw a square.

Two states are used,  **FORWARD** for linear movement and **TURN** for angular movement, the error is checked by the **Control** function and when the error is close to 0, the **SetVariable** flag is changed so new Setpoints can be set.

The roslaunch file contains the launch to the controller node and the path node.

### CODE REVIEW

File: ca_tools/scripts/pos_controller_pid.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+import rospy
+import math
+from nav_msgs.msg import Odometry
+from geometry_msgs.msg import Twist
+from tf.transformations import euler_from_quaternion
+
+
+
+class Execute_Draw_Square:
+
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py:20]: Add a docstring indicating the purpose of this class.
Also please follow the [PEP8](https://www.python.org/dev/peps/pep-0008/#class-names) naming convention.

File: ca_tools/scripts/pos_controller_pid.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    # Time variables
+    now=float
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: why is this needed? can you initialize to a value?

File: ca_tools/scripts/pos_controller_pid.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    timeChange=float
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: same as above

File: ca_tools/scripts/pos_controller_pid.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    Current_length=0.0
+    Setpoint_length=0.0
+    Input_length=0.0
+    error_length=0.0
+    errSum_length=0.0
+    Max_output_length=0.0
+    Min_output_length=0.0
+
+    # PID variables
+    kp=float
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: why are you doing this? (and below)

File: ca_tools/scripts/pos_controller_pid.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    ki=float
+    kd=float
+
+    #State Variables
+    Setpoint_variable=bool
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: Initialize to values, not types. Also follow the PEP8 naming convention.

File: ca_tools/scripts/pos_controller_pid.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    #Twist Variable
+    twist = Twist()
+
+    #Odometry Flag
+    Odom_flag=False
+
+    def __init__(self):
+
+        # PID variables
+        self.kp=1
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: You don't need to "declare" the variables if you define them in the constructor.

File: ca_tools/scripts/pos_controller_pid.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        self.state="FIRST"
+
+        self.Max_output_angle = 0.4
+        self.Min_output_angle = -0.4
+
+        self.Max_output_length = 0.4
+        self.Min_output_length = -0.4
+
+    def callback(self,data):
+        odom_pos = data
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: why is this needed?

File: ca_tools/scripts/pos_controller_pid.py [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        q = (odom_pos.pose.pose.orientation.x,
+             odom_pos.pose.pose.orientation.y,
+             odom_pos.pose.pose.orientation.z,
+             odom_pos.pose.pose.orientation.w)
+        euler_angles = euler_from_quaternion(q)
+        self.Current_angle = euler_angles[2]
+        if self.Current_angle < 0:
+            self.Current_angle= self.Current_angle + 2 * math.pi
+        self.Current_length = math.sqrt((odom_pos.pose.pose.position.x ** 2) + (odom_pos.pose.pose.position.y ** 2))
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: consider using https://docs.python.org/3/library/math.html#math.hypot

**Review by ernestmc — CHANGES_REQUESTED** [review]

**Comment by eborghi10** [comment]
> WORK IN PROGRESS
> 

Add [WIP] in the title of the PR.

File: ca_description/urdf/color_sensor.xacro [hunk 9]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" ?>
```
Comment (eborghi10) [inline ca_description/urdf/color_sensor.xacro:1]: Change the target of this PR and instead of RoboticaUtnFrba:kinetic-devel, target to your PR where you added the color sensor.

**Comment by eborghi10** [comment]
How do I test the PR? Please, add instructions in the description.

File: ca_tools/scripts/pos_controller_pid.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+import rospy
+import math
+from nav_msgs.msg import Odometry
+from geometry_msgs.msg import Twist
+from tf.transformations import euler_from_quaternion
+
+
+
+class Execute_Draw_Square:
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Don't mix CamelCase with snake_case. For class names use CamelCase.

File: ca_tools/scripts/pos_controller_pid.py (hunk 2 above)
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Variables that are used throughout the class, need to have `self.` appended. Check [here](https://medium.com/quick-code/understanding-self-in-python-a3704319e5f0).

File: ca_tools/scripts/pos_controller_pid.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        self.Max_output_length = 0.4
+        self.Min_output_length = -0.4
+
+    def callback(self,data):
+        odom_pos = data
+        q = (odom_pos.pose.pose.orientation.x,
+             odom_pos.pose.pose.orientation.y,
+             odom_pos.pose.pose.orientation.z,
+             odom_pos.pose.pose.orientation.w)
+        euler_angles = euler_from_quaternion(q)
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: You can do this:
```Python
euler_angles = euler_from_quaternion(q)
self.Current_angle = euler_angles[2]
```
in one line using this:
```Python
_, _, self.Current_angle = euler_from_quaternion(q)
```

Comment (eborghi10) [reply ca_tools/scripts/pos_controller_pid.py]: Also, names of the variables are wrong because you're mixing CamelCase with snake_case.

File: ca_tools/scripts/pos_controller_pid.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        self.errSum_length = self.errSum_length + (self.error_length * self.timeChange)
+
+        # rospy.loginfo(Input_angle)
+        # der_error=(error - lastErr)/timeChange
+
+
+    def compute_pid_output(self):
+        # /*Compute PID Output*/
+
+        # Angle Calculations
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Are you doing the same in both conditions? (turn and forward)

File: ca_tools/scripts/pos_controller_pid.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        #twist_publisher.publish(twist)
+
+
+
+    def update_time(self):
+        # lastErr=error
+        if self.now > 0:
+            self.lastTime = self.now
+
+    def set_objectives(self):
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Call it "goals" instead of "objectives"

File: ca_tools/scripts/pos_controller_pid.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+        if self.state=="FIRST":
+            self.Input_lenght=self.Current_length
+            self.Input_angle=self.Current_angle
+            self.Setpoint_length=self.Input_lenght + 2.0
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: What is "+ 2.0" for?

File: ca_tools/scripts/pos_controller_pid.py [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+            self.Setpoint_angle=self.Input_angle
+            self.state="FORWARD"
+            self.Set_variable=False
+            pass
+
+        if self.state=="TURN":
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: I'm seeing a lot of duplicated code. Consider using functions.

File: ca_tools/scripts/pos_controller_pid.py [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+                self.Setpoint_angle = self.Input_angle
+                self.Set_variable=False
+            else:
+                pass
+
+    def control_objectives(self,twist_publisher):
+        #rospy.loginfo(self.error_length)
+        if self.state=="FIRST":
+            pass
+        if self.state=="FORWARD":
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Here you have some duplicated code too.

Comment (eborghi10) [reply ca_tools/scripts/pos_controller_pid.py]: It will be very hard for you to debug if you code like this. Follow the KISS principle (**keep it simple!**).

File: ca_tools/scripts/pos_controller_pid.py [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+                self.Set_variable=True
+                pass
+            else:
+                twist_publisher.publish(self.twist)
+                pass
+
+
+
+def __Execution__():
+    rospy.init_node('control_robot')
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Initialize the ROS node inside the class.

Comment (eborghi10) [reply ca_tools/scripts/pos_controller_pid.py]: And the rest of the logic too.

File: ca_tools/scripts/pos_controller_pid.py [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+            Robot.get_time()
+            Robot.set_objectives()
+            Robot.compute_error()
+            Robot.compute_pid_output()
+            Robot.control_objectives(twist_publisher)
+            Robot.update_time()
+        else:
+            pass
+    rate.sleep()
+    rospy.spin()
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: This is wrong here. Is the callback being called?

**Review by eborghi10 — CHANGES_REQUESTED** [review]
I made some comments, PTAL.

File: ca_description/urdf/color_sensor.xacro (hunk 9 above)
Comment (stevendes) [reply ca_description/urdf/color_sensor.xacro:1]: I'm not sure how to do this, I can change it with the other branches of the rep but I don't know how to do it with the PR

Comment (ernestmc) [reply ca_description/urdf/color_sensor.xacro:1]: You need to target the branch from which you created a PR. Every PR goes from one branch to another. See for example the screenshot below:
image

File: ca_tools/scripts/pos_controller_pid.py (hunk 1 above)
Comment (ernestmc) [reply ca_tools/scripts/pos_controller_pid.py:20]: This is not addressed

File: ca_tools/scripts/pos_controller_pid.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+from nav_msgs.msg import Odometry
+from geometry_msgs.msg import Twist
+from tf.transformations import euler_from_quaternion
+from geometry_msgs.msg import Pose2D
+
+
+class DrawSquare:
+
+    # Time variables
+    Now=0.0
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: Variable names are not following PEP8 convention

File: ca_tools/scripts/pos_controller_pid.py (hunk 5 above)
Comment (ernestmc) [reply ca_tools/scripts/pos_controller_pid.py]: Still not following PEP8 naming convention. Please read the documentation.

File: ca_tools/scripts/pos_controller_pid.py [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+    InputLength=0.0
+    ErrorLength=0.0
+    ErrSumLength=0.0
+    MaxOutputLength=0.0
+    MinOutputLength=0.0
+
+    # PID variables
+    Kp=0
+    Ki=0
+    Kd=0
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: These lines are not needed. You are already defining the variables inside the constructor, lines 68-70

File: ca_tools/scripts/pos_controller_pid.py [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+            self.LastTime = self.Now
+        self.TimeChange = self.Now - self.LastTime
+
+    def update_time(self):
+        if self.Now > 0:
+            self.LastTime = self.Now
+
+    def SetGoals(self):
+
+        if self.State=="FORWARD" and self.SetVariable==True:
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: PEP8 convention requires spaces between expressions. See https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements

File: ca_tools/scripts/pos_controller_pid.py [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+
+    def PID(self,twist_publisher):
+
+        # Error calculations
+
+        self.ErrorAngle=self.SetpointAngle-self.CurrentAngle
+        self.ErrSumAngle+=self.ErrorAngle*self.TimeChange
+        self.ErrorLengthAux.x=self.SetpointLength.x-self.CurrAux.x
+        self.ErrorLengthAux.y=self.SetpointLength.y - self.CurrAux.y
+        self.ErrorLength=self.ErrorLengthAux.x + self.ErrorLengthAux.y
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: This doesn't look right. Why are you adding up X and Y? Please explain

File: ca_tools/scripts/path.py [hunk 23]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
+import rospy
+
+from nav_msgs.msg import Path
```
Comment (eborghi10) [inline ca_tools/scripts/path.py]: You can merge these 2 lines together:

```Python
from nav_msgs.msg import Path, Odometry
```

File: ca_tools/scripts/path.py [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+from nav_msgs.msg import Odometry
+from geometry_msgs.msg import PoseStamped
+
+path = Path()
```
Comment (eborghi10) [inline ca_tools/scripts/path.py]: Move all this code into a new class declared here.

**Comment by ernestmc** [comment]
- Need to work on complying with the PEP8 standard, specially name of variables and spacing in expressions.
- Add docstring comments to all classes and methods
- Pay attention to the other comments I made

File: ca_tools/scripts/pos_controller_pid.py [hunk 25]
```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python
+
+import rospy
+import math
+from nav_msgs.msg import Odometry
+from geometry_msgs.msg import Twist
+from tf.transformations import euler_from_quaternion
+from geometry_msgs.msg import Pose2D
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Use only one include of `geometry_msgs`

File: ca_tools/scripts/pos_controller_pid.py [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+    Kd=0
+
+    # Auxiliar Variables
+    CurrAux = Pose2D()
+    IniAux=Pose2D()
+    FinAux=Pose2D()
+    ErrorLengthAux=Pose2D()
+
+    #State Variables
+    #Set Variable is TRUE if we need to set new goals
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Isn't better to call it `is_new_goal`?

File: ca_tools/scripts/pos_controller_pid.py (hunk 19 above)
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: I recommend you to declare here variables that are not going to be modified throughout the code. And in `__init__` variables that need to be initialized.

File: ca_tools/scripts/pos_controller_pid.py [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+            data.pose.pose.orientation.y,
+            data.pose.pose.orientation.z,
+            data.pose.pose.orientation.w)
+
+        _, _, self.CurrentAngle = euler_from_quaternion(q)
+
+        self.CurrAux.x=data.pose.pose.position.x
+        self.CurrAux.y=data.pose.pose.position.y
+
+        self.OdomFlag+=1
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Two things regarding this flag:

1. If the value will be between 0 and 1, why don't you use a boolean?
2. What happens if you remove the flag and the update loop updates your math every time? Is there an issue with that?

File: ca_tools/scripts/pos_controller_pid.py [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+
+    def get_time(self):
+        self.Now = rospy.get_time()
+        if self.LastTime < 0.001:
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: This check here is not clear to me and I don't think it's correct.
What are you trying to do here?

Comment (eborghi10) [reply ca_tools/scripts/pos_controller_pid.py]: Ok, it seems that you're working with complex logic. Let's discuss this during the meeting.

Comment (eborghi10) [reply ca_tools/scripts/pos_controller_pid.py]: Gazebo plugins will give you an idea of how to work with time and update rates in code.

File: ca_tools/scripts/pos_controller_pid.py (hunk 21 above)
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: You don't need to do comparations with `==True`. Just use `if self._set_variable:`

File: ca_tools/scripts/pos_controller_pid.py [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+            self.SetpointLength=self.FinAux
+            self.SetpointAngle=self.InputAngle
+            self.SetVariable=False
+
+        if self.State=="TURN" and self.SetVariable==True:
+            self.InputLength=self.IniAux
+            self.InputAngle=self.CurrentAngle
+            self.SetpointLength=self.InputLength
+            self.SetpointAngle=self.InputAngle + math.pi/2
+            if self.SetpointAngle>math.pi:
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Call a method that wraps angles instead.

File: ca_tools/scripts/pos_controller_pid.py (hunk 22 above)
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Errors mustn't be calculated like this! Why did you remove `hypot`?

File: ca_tools/scripts/pos_controller_pid.py [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+        if self.Output_angle>self.MaxOutputAngle or self.Output_angle<0:
+            self.Output_angle=self.MaxOutputAngle
+
+        self.Vel.linear.x=self.Output_length
+        self.Vel.angular.z=self.Output_angle
+        twist_publisher.publish(self.Vel)
+
+    def Control(self):
+        if self.State=="FORWARD" and self.SetVariable==False:
+            if abs(self.ErrorLength)<00.1:
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: Typo: `00.1`

File: ca_tools/scripts/pos_controller_pid.py [hunk 31]
```diff
@@ -0,0 +1,190 @@
...
+                self.State="TURN"
+                self.SetVariable=True
+
+        if self.State=="TURN" and self.SetVariable==False:
+            if abs(self.ErrorAngle)<0.01:
```
Comment (eborghi10) [inline ca_tools/scripts/pos_controller_pid.py]: `0.001` must be a class constant, not a hardcoded value here.

**Review by eborghi10 — CHANGES_REQUESTED** [review]

**Comment by stevendes** [comment]
**UPDATE**

Still have some issues to address, not necessary to review it now.

Some issues like the PEP8 format addressed and some of the recommendations.

File: ca_tools/scripts/pos_controller_pid.py [hunk 32]
```diff
@@ -0,0 +1,212 @@
...
+        self._set_angle = 0.0
+        self._input_angle = 0.0
+        self._error_angle = 0.0
+        self._err_sum_angle = 0.0
+        self._output_angle = 0.0
+        self._control_error_angle = 0.01
+
+        # Lenght variables
+
+        self._set_length = Pose2D()
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: Why is the length a pose? That doesn't make sense. Probably the variable needs to be named differently

File: ca_tools/scripts/pos_controller_pid.py [hunk 33]
```diff
@@ -0,0 +1,212 @@
...
+        self._input_length = 0.0
+        self._error_length = 0.0
+        self._err_sum_length = 0.0
+        self._output_length = 0.0
+        self._control_error_length = 0.1
+
+        # Auxiliar Variables
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: Do all these variables really need to be global to the class?

File: ca_tools/scripts/pos_controller_pid.py [hunk 34]
```diff
@@ -0,0 +1,212 @@
...
+        if self._now > 0:
+            self._last_time = self._now
+
+    def set_goals(self):
+        """
+        Function that sets the new goals checking the state and if the _set_variable flag
+        is turned True, if it is true, means that a new goal needs to be set.
+        """
+
+
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py:107]: extra line

File: ca_tools/scripts/pos_controller_pid.py [hunk 35]
```diff
@@ -0,0 +1,212 @@
...
+        Function to update the Last Time.
+        """
+
+        if self._now > 0:
+            self._last_time = self._now
+
+    def set_goals(self):
+        """
+        Function that sets the new goals checking the state and if the _set_variable flag
+        is turned True, if it is true, means that a new goal needs to be set.
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py:103]: Wouldn't it be better that this function returns the boolean value, instead of setting global variables?

File: ca_tools/scripts/pos_controller_pid.py [hunk 36]
```diff
@@ -0,0 +1,212 @@
...
+            self._output_length = 0.0
+
+        if self._output_length > self._max_output_length:
+            self._output_length = self._max_output_length
+        if self._output_length < 0:
+            self._output_length = self._max_output_length
+        if self._output_angle > self._max_output_angle or self._output_angle < 0:
+            self._output_angle = self._max_output_angle
+
+        self._vel.linear.x = self._output_length
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: Beware that this is a speed measured in `m/s` and your are assigning a length

File: ca_tools/scripts/pos_controller_pid.py [hunk 37]
```diff
@@ -0,0 +1,212 @@
...
+        self._vel.angular.z = self._output_angle
```
Comment (ernestmc) [inline ca_tools/scripts/pos_controller_pid.py]: Beware this is an angular speed measured in `rad/s` and you are assigning an angle.
It is not completely wrong, but you should be careful that the values are not the same. Specially in the case of angular speed, it doesn't make sense to clip an angle, because the robot can turn as fast as it goes.

**Review by ernestmc — CHANGES_REQUESTED** [review]
Good work @stevendes! This is looking much better! I left a couple more comments.

File: ca_tools/scripts/pos_controller_pid.py (hunk 21 above)
Comment (stevendes) [reply ca_tools/scripts/pos_controller_pid.py]: This is not the latest version, I'm not sure why this appears to you.

File: ca_tools/scripts/pos_controller_pid.py (hunk 22 above)
Comment (stevendes) [reply ca_tools/scripts/pos_controller_pid.py]: This too, I already change it in the latest version, not sure why still appears.

File: ca_tools/scripts/pos_controller_pid.py (hunk 20 above)
Comment (stevendes) [reply ca_tools/scripts/pos_controller_pid.py]: This too is already addressed in the latest version

File: ca_tools/scripts/pos_controller_pid.py (hunk 35 above)
Comment (stevendes) [reply ca_tools/scripts/pos_controller_pid.py:103]: I'm not sure how to approach this.

Comment (ernestmc) [reply ca_tools/scripts/pos_controller_pid.py:103]: Nevermind, it's good enough. In general we should try to avoid using too many global variables. Use more parameters and return values from the functions, when possible.

**Review by ernestmc — APPROVED** [review]
lgtm

**Comment by eborghi10** [comment]
> > WORK IN PROGRESS
> 
> Add [WIP] in the title of the PR.

@stevendes Is this a WIP PR?

**Review by eborghi10 — APPROVED** [review]

**Comment by stevendes** [comment]
> > > WORK IN PROGRESS
> > 
> > 
> > Add [WIP] in the title of the PR.
> 
> @stevendes Is this a WIP PR?

I forgot to remote the tag, fixed!

**Comment by eborghi10** [comment]
Closing it because this was intended to be used as training.
